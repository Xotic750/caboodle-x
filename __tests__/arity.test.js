import {arity} from '../dist/caboodle-x';
import primitives from './helpers/primitives';
import objects from './helpers/objects';

const noop = function () {};

describe('arity', () => {
  // let actual;

  /* istanbul ignore next */
  const testSubject = {
    push(o) {
      this.a.push(o);
    },
  };

  beforeEach(() => {
    // actual = [];
    testSubject.a = [];
  });

  it('is a function', () => {
    expect(typeof arity).toBe('function');
  });

  it('should throw when target is null or undefined', () => {
    expect(() => {
      arity();
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      arity(undefined);
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      arity(null);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw when target is not a function', () => {
    primitives.forEach((primitive) => {
      expect(() => {
        arity(primitive);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  it('binds properly without arguments and not called with arguments', () => {
    let context;
    let args;
    const fn = function _fn(a, b, c) {
      context = this;
      args = Array.prototype.slice.call(arguments);
      noop(a, b, c);
    };

    testSubject.func = arity(fn);
    expect(typeof testSubject.func).toBe('function');
    testSubject.func();

    expect(context).toBe(testSubject);
    expect(args).toEqual([]);
  });

  it('binds properly without arguments', () => {
    let context;
    let args;
    const fn = function _fn(a, b, c) {
      context = this;
      args = Array.prototype.slice.call(arguments);
      noop(a, b, c);
    };

    testSubject.func = arity(fn);
    expect(typeof testSubject.func).toBe('function');
    testSubject.func(1, 2, 3, 4);

    expect(context).toBe(testSubject);
    expect(args).toEqual([1, 2, 3]);
  });

  it('binds properly with arguments', () => {
    const arityLength = 1;
    let context;
    let args;
    const fn = function _fn(a, b, c) {
      context = this;
      args = Array.prototype.slice.call(arguments);
      noop(a, b, c);
    };

    testSubject.func = arity(fn, arityLength);
    expect(typeof testSubject.func).toBe('function');
    testSubject.func(1, 2, 3, 4);

    expect(context).toBe(testSubject);
    expect(args).toEqual([1]);
  });

  it("has the new instance's context as a constructor", () => {
    let context;
    let args;
    const fn = function _fn(a, b, c) {
      context = this;
      args = Array.prototype.slice.call(arguments);
      noop(a, b, c);
    };

    testSubject.Func = arity(fn);
    const result = new testSubject.Func();
    expect(result).toBeTruthy();
    expect(context instanceof fn).toBe(true);
    expect(result instanceof fn).toBe(true);
    expect(args).toEqual([]);
  });

  it("has the new instance's context as a constructor with arity set", () => {
    let context;
    let args;
    const fn = function _fn(a, b, c) {
      context = this;
      args = Array.prototype.slice.call(arguments);
      noop(a, b, c);
    };

    testSubject.Func = arity(fn, 2);
    const result = new testSubject.Func(1, 2, 3);
    expect(result).toBeTruthy();
    expect(context instanceof fn).toBe(true);
    expect(result instanceof fn).toBe(true);
    expect(args).toEqual([1, 2]);
  });

  it('passes the correct arguments as a constructor', () => {
    const expected = {name: 'Correct'};
    const fn = function _fn(arg) {
      expect(Object.prototype.hasOwnProperty.call(this, 'name')).toBe(false);
      return arg;
    };

    testSubject.Func = arity(fn);
    const ret = new testSubject.Func(expected);
    expect(ret).toBe(expected);
  });

  it('returns the return value of the bound function when called as a constructor', () => {
    const oracle = [1, 2, 3];
    const fn = function _fn() {
      expect(this).not.toBe(oracle);
      return oracle;
    };

    const Subject = arity(fn);
    const result = new Subject();
    expect(result).toBe(oracle);
  });

  it('returns the correct value if constructor returns primitive', () => {
    const fn = function _fn(oracle) {
      expect(this).not.toBe(oracle);
      return oracle;
    };

    const Subject = arity(fn);

    primitives.forEach((primitive) => {
      expect(new Subject(primitive)).not.toBe(primitive);
    });

    objects.forEach((object) => {
      expect(new Subject(object)).toBe(object);
    });
  });

  it('returns the value that instance of original "class" when called as a constructor', () => {
    /* istanbul ignore next */
    const ClassA = function (x) {
      this.name = x || 'A';
    };

    const ClassB = arity(ClassA, 'B');

    const result = new ClassB();
    expect(result instanceof ClassA).toBe(true);
    expect(result instanceof ClassB).toBe(true);
  });
});
