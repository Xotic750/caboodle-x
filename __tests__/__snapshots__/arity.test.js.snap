// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arity should throw when target is not a function 1`] = `"Not a function"`;

exports[`arity should throw when target is not a function 2`] = `"Not a function"`;

exports[`arity should throw when target is not a function 3`] = `"Not a function"`;

exports[`arity should throw when target is not a function 4`] = `"Not a function"`;

exports[`arity should throw when target is not a function 5`] = `"Not a function"`;

exports[`arity should throw when target is not a function 6`] = `"Not a function"`;

exports[`arity should throw when target is not a function 7`] = `"Not a function"`;

exports[`arity should throw when target is not a function 8`] = `"Not a function"`;

exports[`arity should throw when target is not a function 9`] = `"Not a function"`;

exports[`arity should throw when target is not a function 10`] = `"Not a function"`;

exports[`arity should throw when target is not a function 11`] = `"Not a function"`;

exports[`arity should throw when target is not a function 12`] = `"Not a function"`;

exports[`arity should throw when target is not a function 13`] = `"Not a function"`;

exports[`arity should throw when target is not a function 14`] = `"Not a function"`;

exports[`arity should throw when target is not a function 15`] = `"Not a function"`;

exports[`arity should throw when target is not a function 16`] = `"Not a function"`;

exports[`arity should throw when target is not a function 17`] = `"Not a function"`;

exports[`arity should throw when target is not a function 18`] = `"Not a function"`;

exports[`arity should throw when target is not a function 19`] = `"Not a function"`;

exports[`arity should throw when target is not a function 20`] = `"Not a function"`;

exports[`arity should throw when target is not a function 21`] = `"Not a function"`;

exports[`arity should throw when target is not a function 22`] = `"Not a function"`;

exports[`arity should throw when target is not a function 23`] = `"Not a function"`;

exports[`arity should throw when target is not a function 24`] = `"Not a function"`;

exports[`arity should throw when target is not a function 25`] = `"Not a function"`;

exports[`arity should throw when target is not a function 26`] = `"Not a function"`;

exports[`arity should throw when target is not a function 27`] = `"Not a function"`;

exports[`arity should throw when target is not a function 28`] = `"Not a function"`;

exports[`arity should throw when target is not a function 29`] = `"Not a function"`;

exports[`arity should throw when target is not a function 30`] = `"Not a function"`;

exports[`arity should throw when target is not a function 31`] = `"Not a function"`;

exports[`arity should throw when target is not a function 32`] = `"Not a function"`;

exports[`arity should throw when target is not a function 33`] = `"Not a function"`;

exports[`arity should throw when target is not a function 34`] = `"Not a function"`;

exports[`arity should throw when target is not a function 35`] = `"Not a function"`;

exports[`arity should throw when target is not a function 36`] = `"Not a function"`;

exports[`arity should throw when target is not a function 37`] = `"Not a function"`;

exports[`arity should throw when target is not a function 38`] = `"Not a function"`;

exports[`arity should throw when target is not a function 39`] = `"Not a function"`;

exports[`arity should throw when target is not a function 40`] = `"Not a function"`;

exports[`arity should throw when target is not a function 41`] = `"Not a function"`;

exports[`arity should throw when target is not a function 42`] = `"Not a function"`;

exports[`arity should throw when target is not a function 43`] = `"Not a function"`;

exports[`arity should throw when target is not a function 44`] = `"Not a function"`;

exports[`arity should throw when target is null or undefined 1`] = `"Not a function"`;

exports[`arity should throw when target is null or undefined 2`] = `"Not a function"`;

exports[`arity should throw when target is null or undefined 3`] = `"Not a function"`;
